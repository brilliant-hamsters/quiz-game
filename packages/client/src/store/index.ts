import { combineReducers, configureStore } from '@reduxjs/toolkit';
import authReducer from './auth/authSlice';
import profileReducer from './profile/profileSlice';
import leaderboardReducer from './leaderboard/leaderboardSlice';
import forumReducer from './forum/forumSlice'
import { useDispatch, useSelector, TypedUseSelectorHook } from 'react-redux';
import { persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

const reducers = combineReducers({
  auth: authReducer,
  profile: profileReducer,
  leaderboard: leaderboardReducer,
  forum: forumReducer,
})

const persistConfig = {
  key: 'root',
  whitelist: ['auth', 'profile', 'leaderboard'],
  storage
};

const persistedReducer = persistReducer(persistConfig, reducers)

let store: any

export const createStore = (initialStore?: any) => {
  store = configureStore({
    preloadedState: initialStore,
    reducer: persistedReducer,
    middleware: getDefaultMiddleware => {
      return getDefaultMiddleware({
        serializableCheck: false,
      })
    },
  })
  return store
}

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch

export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
