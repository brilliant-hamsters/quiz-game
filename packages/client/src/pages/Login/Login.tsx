import { ChangeEvent, useEffect, useState } from 'react'
import { useNavigate } from 'react-router-dom'
import { Input } from '../../components/Input'
import { PageWithForm } from '../../components/PageWithForm/PageWithForm'
import { useAppDispatch, useAppSelector } from '../../store'
import { signIn } from '../../store/auth/authSlice'
import styles from './login.module.scss'
//import { useValidation } from './../../utils/hooks/validation.hook'
//import Api from '../../api/ApiBase'

export function Login() {
  //const [validObj, setValidity] = useValidation()
  const [dataLogin, onChangeDataLogin] = useState<DataAuth>({
    login: '',
    password: '',
  })
  const dispatch = useAppDispatch()
  const { loggedIn, isLoading } = useAppSelector(state => state.auth)
  const navigate = useNavigate()

  useEffect(() => {
    if (loggedIn === true) navigate('/')
  }, [loggedIn])

  function changeData(e: ChangeEvent<HTMLInputElement>) {
    const value = e.target.value
    const nameInput = e.target.name
    onChangeDataLogin({ ...dataLogin, [nameInput]: value })
    // setValidity(e)
  }

  function onSubmitForm() {
    dispatch(signIn(dataLogin))
  }

  return (
    <main className={styles.main}>
      <PageWithForm
        classElement="login"
        title="Авторизация"
        buttonName="Войти"
        isFormValid
        caption="Нет аккаунта?"
        path="/sign_up"
        linkName="Зарегистрироваться"
        onSubmitForm={onSubmitForm}
        isLoading={isLoading}>
        <Input
          classInput="login"
          type="text"
          name="login"
          autoFocus
          required
          label="Логин"
          onChange={changeData}
          value={dataLogin.login}
          // isValid={validObj.login?.valid}
          // errorMessage={validObj.login?.text}
        />
        <Input
          classInput="login"
          type="password"
          name="password"
          autoFocus={false}
          required
          label="Пароль"
          onChange={changeData}
          value={dataLogin.password}
          // isValid={validObj.password?.valid}
          // errorMessage={validObj.password?.text}
        />
      </PageWithForm>
    </main>
  )
}
