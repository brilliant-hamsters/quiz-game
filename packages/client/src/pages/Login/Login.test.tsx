import React from 'react';
import { render, screen} from "@testing-library/react"
import Login  from './Login';
import { Provider } from 'react-redux';
import store from '../../store';
import '@testing-library/jest-dom'
import useEvent from "@testing-library/user-event"
import { BrowserRouter, Route, Routes } from 'react-router-dom';

describe('Компонент Логин', () => {
  const {asFragment} = render(
    <Provider store={store}>
      <BrowserRouter>
          <Routes>
            <Route path="/" element={<Login />}></Route>
        </Routes>
      </BrowserRouter>
    </Provider>
  )

  test('-> Input must be in the login ', () => {
      expect(screen.getByLabelText<HTMLSelectElement>('Логин')).toBeInTheDocument();
  })

  test('-> Changing an input when entering a value', async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
            <Routes>
              <Route path="/" element={<Login />}></Route>
          </Routes>
        </BrowserRouter>
      </Provider>
    )
     await useEvent.type(screen.getByLabelText<HTMLInputElement>('Логин'), "TestID");
     await useEvent.type(screen.getByLabelText<HTMLInputElement>('Пароль'), "TestPassword");

    expect(screen.getByLabelText<HTMLSelectElement>(/Логин/i).value).toBe("TestID");
    expect(screen.getByLabelText<HTMLSelectElement>(/Пароль/i).value).toBe("TestPassword")
  });

  test('-> Login snapshot', async () => {
    expect(asFragment()).toMatchSnapshot();
  });
})


