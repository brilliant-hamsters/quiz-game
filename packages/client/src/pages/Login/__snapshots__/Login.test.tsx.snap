// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Компонент Логин -> Login snapshot 1`] = `
<DocumentFragment>
  <main>
    <div
      class="undefined undefined"
    >
      <div />
      <h1>
        Авторизация
      </h1>
      <form
        class="undefined undefined"
        novalidate=""
      >
        <fieldset>
          <label
            class="undefined undefined"
          >
            Логин
            <input
              id="input-login"
              name="login"
              required=""
              type="text"
              value=""
            />
          </label>
          <label
            class="undefined undefined"
          >
            Пароль
            <input
              id="input-password"
              name="password"
              required=""
              type="password"
              value=""
            />
          </label>
        </fieldset>
        <div>
          <button
            class="undefined undefined"
            type="submit"
          >
            Войти
          </button>
          <span>
            или
          </span>
          <button
            class="undefined undefined"
            type="submit"
          >
            Войти с
            <div />
          </button>
        </div>
        <span>
          Нет аккаунта?
          <a
            href="/sign_up"
          >
            Зарегистрироваться
          </a>
        </span>
      </form>
    </div>
  </main>
</DocumentFragment>
`;
